src/machine/machine_rp2040-rtc.go

```diff
@@ -0,0 +1,256 @@
+//go:build rp2040
+
+// Implementation based on code located here:
+// https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_rtc/rtc.c
+
+package machine
+
+import (
+	"device/rp"
+	"errors"
+	"runtime/interrupt"
+	"runtime/volatile"
+	"unsafe"
+)
+
+type rtcType struct {
+	clkDivM1  volatile.Register32
+	setup0    volatile.Register32
+	setup1    volatile.Register32
+	ctrl      volatile.Register32
+	irqSetup0 volatile.Register32
+	irqSetup1 volatile.Register32
+	rtc1      volatile.Register32
+	rtc0      volatile.Register32
+	intR      volatile.Register32
+	intE      volatile.Register32
+	intF      volatile.Register32
+	intS      volatile.Register32
+}
+
+var RTC = (*rtcType)(unsafe.Pointer(rp.RTC))
+
+var rtcAlarmRepeats bool
+var rtcCallback func()
+
+var ErrRtcNotRunning = errors.New("RTC not running")
+var ErrRtcInvalidTime = errors.New("invalid time for RTC")
+
+func (rtc *rtcType) running() bool {
+	return rtc.ctrl.HasBits(rp.RTC_CTRL_RTC_ACTIVE)
+}
+
+func (rtc *rtcType) init() {
+	// Get clk_rtc freq and make sure it is running
+	rtcFreq := configuredFreq[clkRTC]
+	if rtcFreq == 0 {
+		panic("rtc freq is zero")
+	}
+
+	// Take rtc out of reset now that we know clk_rtc is running
+	resetBlock(rp.RESETS_RESET_RTC)
+	unresetBlockWait(rp.RESETS_RESET_RTC)
+
+	// Set up the 1 second divider.
+	// If rtc_freq is 400 then clkdiv_m1 should be 399
+	rtcFreq -= 1
+
+	// Check the freq is not too big to divide
+	if rtcFreq > rp.RTC_CLKDIV_M1_CLKDIV_M1_Msk {
+		panic("rtc freq is too big to divide")
+	}
+
+	// Write divide value
+	rtc.clkDivM1.Set(rtcFreq)
+}
+
+func (rtc *rtcType) SetTime(t RtcTime) error {
```

Comment:
I'd _love_ to. Unfortunately, that makes a cycle:
```
package github.com/tinygo-org/tinygo/src/examples/rtc
        imports machine
        imports time
        imports runtime
        imports machine: import cycle not allowed
```

Reply:
I'd go for a `SetDate` function that follows the [stdlib format of creating a new date](https://pkg.go.dev/time#Date) (excluding the loc).
This would probably make it more natural to use than with a machine specific type.

Reply:
Check this out, it might be a way of solving this cyclical dependency problem given that we all agree it's time the machine package had it's own Time type.

`machine/time.go`
```go
type Time struct {
 // Same fields as time.Time
}

func (rtc *RTC) SetTime[T ~Time](t T) error {
   // do the setting.
}
```

`time/time.go`
```go
// replace contents of time.Time type.
type Time machine.Time
```
I have no idea of the far reaching consequences of this change.

Reply:
I've actually made an alternative PR #3403 that sets RTC interrupt with a delay instead of actual time. Probably, that's better than such tricks.